File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from sys import stdin

def main():
    n, m, v = map(int, stdin.readline().split())
    matrix = [[0] * (n + 1) for _ in range(n + 1)]
    for _ in range(m):
        a, b = map(int, stdin.readline().split())
        matrix[a][b] = matrix[b][a] = 1

    def dfs(start, visited):
        visited += [start]
        for c in range(len(matrix[start])):
            if matrix[start][c] == 1 and (c not in visited):
                dfs(c, visited)
        return visited

    def bfs(start):
        visited = [start]
        queue = [start]
        while queue:
            n = queue.pop(0)
            for c in range(len(matrix[start])):
                if matrix[n][c] == 1 and (c not in visited):
                    visited.append(c)
                    queue.append(c)
        return visited

    print(*dfs(v,[]))
    print(*bfs(v))

main()

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

	static int n, m, v;
	static HashMap<Integer, List<Integer>> graph;
	static StringBuilder sb;
	static boolean[] visited;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		v = Integer.parseInt(st.nextToken());

		graph = new HashMap<>();
		for (int i = 1; i < n + 1; i++) {
			graph.put(i, new ArrayList<>());
		}

		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			graph.get(a).add(b);
			graph.get(b).add(a);
		}

		for (int i = 1; i < n + 1; i++) {
			Collections.sort(graph.get(i));
		}

		sb = new StringBuilder();

		visited = new boolean[n + 1];
		visited[v] = true;
		dfs(v);
		sb.append("\n");
		visited = new boolean[n + 1];
		bfs(v);

		System.out.println(sb);
	}

	private static void dfs(int curr) {
		sb.append(curr).append(" ");

		for (Integer nextNode : graph.get(curr)) {
			if (!visited[nextNode]) {
				visited[nextNode] = true;
				dfs(nextNode);
			}
		}
	}

	private static void bfs(int start) {
		Queue<Integer> que = new ArrayDeque<>();
		que.offer(start);
		visited[start] = true;

		while (!que.isEmpty()) {
			int node = que.poll();
			sb.append(node).append(" ");
			for (Integer nextNode : graph.get(node)) {
				if (!visited[nextNode]) {
					visited[nextNode] = true;
					que.offer(nextNode);
				}
			}
		}
	}

}